File: prefpin-processor/src/main/java/prefpin/processor/PrefPinProcessor.java
package prefpin.processor;

import com.google.auto.service.AutoService;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.tools.Diagnostic;
import prefpin.BindPref;
import prefpin.OnPrefChange;
import prefpin.OnPrefClick;

@AutoService(Processor.class)
public class PrefPinProcessor extends AbstractProcessor {

  @Override public Set<String> getSupportedAnnotationTypes() {
    Set<String> types = new LinkedHashSet<>();
    types.add(BindPref.class.getCanonicalName());
    types.add(OnPrefClick.class.getCanonicalName());
    types.add(OnPrefChange.class.getCanonicalName());
    return types;
  }

  @Override public boolean process(Set<? extends TypeElement> annotations,
      RoundEnvironment roundEnvironment) {

    Map<TypeElement, Set<Element>> bindingMap = new LinkedHashMap<>();
    parsePreferenceBinding(roundEnvironment, bindingMap, BindPref.class);
    parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefClick.class);
    parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefChange.class);

    if (!bindingMap.isEmpty()) {
      for (Map.Entry<TypeElement, Set<Element>> entry : bindingMap.entrySet()) {
        String targetClassName = entry.getKey().getQualifiedName().toString();

        try {
          writeBinding(targetClassName, entry.getValue());
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
    }

    return true;
  }

  private void parsePreferenceBinding(RoundEnvironment roundEnv,
      Map<TypeElement, Set<Element>> bindingMap, Class<? extends Annotation> annotation) {
    for (Element element : roundEnv.getElementsAnnotatedWith(annotation)) {
      if (element.getModifiers().contains(Modifier.PRIVATE)) {
        processingEnv.getMessager()
            .printMessage(Diagnostic.Kind.ERROR,
                "Binding annotation can not applied to private fields or methods.", element);
      }

      TypeElement targetPrefFragment = (TypeElement) element.getEnclosingElement();
      if (bindingMap.containsKey(targetPrefFragment)) {
        bindingMap.get(targetPrefFragment).add(element);
      } else {
        Set<Element> fields = new LinkedHashSet<>();
        fields.add(element);
        bindingMap.put(targetPrefFragment, fields);
      }
    }
  }

  /**
   * Checks whether a field is annotated correctly with {@link prefpin.BindPref} and prints error
   * message for incorrect one.
   */
  private boolean checkPreferenceAnnotation(Element element) {
    if (isSubtypeOfType(element.asType(), "android.preference.Preference")) {
      return true;
    } else {
      processingEnv.getMessager()
          .printMessage(Diagnostic.Kind.ERROR,
              "@PrefPin must be applied to Preference or its subclass fields", element);
      return false;
    }
  }

  private void writeBinding(String targetClassName, Set<Element> annotationFields)
      throws IOException {
    String packageName = null;
    int lastDot = targetClassName.lastIndexOf('.');
    if (lastDot > 0) {
      packageName = targetClassName.substring(0, lastDot);
    }

    String targetSimpleClassName = targetClassName.substring(lastDot + 1);
    String bindingClassName = targetClassName + "_PrefBinding";
    String bindingSimpleClassName = bindingClassName.substring(lastDot + 1);

    ClassName targetClass = ClassName.get(packageName, targetSimpleClassName);

    FieldSpec targetField = FieldSpec.builder(targetClass, "target", Modifier.PRIVATE).build();

    TypeSpec binding = TypeSpec.classBuilder(bindingSimpleClassName)
        .addModifiers(Modifier.PUBLIC)
        .addField(targetField)
        .addMethod(buildConstructor(targetClass, annotationFields))
        .build();

    JavaFile javaFile = JavaFile.builder(packageName, binding).build();

    javaFile.writeTo(processingEnv.getFiler());
  }

  private MethodSpec buildConstructor(ClassName targetClass, Set<Element> annotationFields) {
    MethodSpec.Builder constructorBuilder = MethodSpec.constructorBuilder()
        .addModifiers(Modifier.PUBLIC)
        .addParameter(targetClass, "target", Modifier.FINAL)
        .addStatement("this.target = target");

    for (Element element : annotationFields) {
      buildFieldBinding(constructorBuilder, element);
      buildOnClickBinding(constructorBuilder, element);
      buildOnChangeBinding(constructorBuilder, element);
    }

    return constructorBuilder.build();
  }

  private void buildFieldBinding(MethodSpec.Builder constructorBuilder, Element element) {
    BindPref bindPref = element.getAnnotation(BindPref.class);
    if (bindPref != null) {
      int resourceId = bindPref.value();

      constructorBuilder.addStatement("target."
          + element.getSimpleName()
          + " = "
          + "("
          + element.asType().toString()
          + ")"
          + "target.findPreference(target.getString("
          + resourceId
          + "))");
    }
  }

  private void buildOnClickBinding(MethodSpec.Builder constructorBuilder, Element element) {
    OnPrefClick onPrefClick = element.getAnnotation(OnPrefClick.class);
    if (onPrefClick != null) {
      int resourceId = onPrefClick.value();

      constructorBuilder.addStatement("target.findPreference(target.getString("
          + resourceId
          + "))"
          + ".setOnPreferenceClickListener(new android.preference.Preference.OnPreferenceClickListener(){\n"
          + "@Override public boolean onPreferenceClick(android.preference.Preference preference) {\n"
          + "\t\ttarget."
          + element.getSimpleName()
          + "(preference);\n"
          + "\t\treturn true;\n"
          + "\t}\n"
          + "})");
    }
  }

  private void buildOnChangeBinding(MethodSpec.Builder constructorBuilder, Element element) {
    OnPrefChange onPrefChange = element.getAnnotation(OnPrefChange.class);
    if (onPrefChange != null) {
      int resourceId = onPrefChange.value();

      constructorBuilder.addStatement("target.findPreference(target.getString("
          + resourceId
          + "))"
          + ".setOnPreferenceChangeListener(new android.preference.Preference.OnPreferenceChangeListener() {\n"
          + "@Override public boolean onPreferenceChange(android.preference.Preference preference, Object newValue) {\n"
          + "\t\ttarget."
          + element.getSimpleName()
          + "(preference, newValue);\n"
          + "\t\treturn true;\n"
          + "\t}\n"
          + "})");
    }
  }

  private boolean isSubtypeOfType(TypeMirror typeMirror, String otherType) {
    if (isTypeEqual(typeMirror, otherType)) {
      return true;
    }
    if (typeMirror.getKind() != TypeKind.DECLARED) {
      return false;
    }
    DeclaredType declaredType = (DeclaredType) typeMirror;
    List<? extends TypeMirror> typeArguments = declaredType.getTypeArguments();
    if (typeArguments.size() > 0) {
      StringBuilder typeString = new StringBuilder(declaredType.asElement().toString());
      typeString.append('<');
      for (int i = 0; i < typeArguments.size(); i++) {
        if (i > 0) {
          typeString.append(',');
        }
        typeString.append('?');
      }
      typeString.append('>');
      if (typeString.toString().equals(otherType)) {
        return true;
      }
    }
    Element element = declaredType.asElement();
    if (!(element instanceof TypeElement)) {
      return false;
    }
    TypeElement typeElement = (TypeElement) element;
    TypeMirror superType = typeElement.getSuperclass();
    if (isSubtypeOfType(superType, otherType)) {
      return true;
    }
    for (TypeMirror interfaceType : typeElement.getInterfaces()) {
      if (isSubtypeOfType(interfaceType, otherType)) {
        return true;
      }
    }
    return false;
  }

  private static boolean isTypeEqual(TypeMirror typeMirror, String otherType) {
    return otherType.equals(typeMirror.toString());
  }
}

File: prefpin-processor/src/main/java/prefpin/processor/PrefPinProcessor.kt
package prefpin.processor

import com.google.auto.service.AutoService
import com.squareup.kotlinpoet.*

import java.io.IOException
import java.util.LinkedHashMap
import java.util.LinkedHashSet

import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.Processor
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror
import javax.tools.Diagnostic

import prefpin.BindPref
import prefpin.BindPrefString
import prefpin.OnPrefChange
import prefpin.OnPrefChangeString
import prefpin.OnPrefClick
import prefpin.OnPrefClickString

@AutoService(Processor::class)
class PrefPinProcessor : AbstractProcessor() {
    override fun getSupportedAnnotationTypes(): Set<String> {
        val types = LinkedHashSet<String>()
        types.add(BindPref::class.java.canonicalName)
        types.add(OnPrefClick::class.java.canonicalName)
        types.add(OnPrefChange::class.java.canonicalName)
        types.add(BindPrefString::class.java.canonicalName)
        types.add(OnPrefClickString::class.java.canonicalName)
        types.add(OnPrefChangeString::class.java.canonicalName)
        return types
    }

    override fun process(annotations: Set<TypeElement>,
                         roundEnvironment: RoundEnvironment): Boolean {

        val bindingMap = LinkedHashMap<TypeElement, LinkedHashSet<Element>>()
        parsePreferenceBinding(roundEnvironment, bindingMap, BindPref::class.java)
        parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefClick::class.java)
        parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefChange::class.java)
        parsePreferenceBinding(roundEnvironment, bindingMap, BindPrefString::class.java)
        parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefClickString::class.java)
        parsePreferenceBinding(roundEnvironment, bindingMap, OnPrefChangeString::class.java)

        if (bindingMap.isNotEmpty()) {
            for ((key, value) in bindingMap) {
                val targetClassName = key.qualifiedName.toString()

                try {
                    writeBinding(targetClassName, value)
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }

        return true
    }

    private fun parsePreferenceBinding(roundEnv: RoundEnvironment,
                                       bindingMap: MutableMap<TypeElement, LinkedHashSet<Element>>, annotation: Class<out Annotation>) {
        for (element in roundEnv.getElementsAnnotatedWith(annotation)) {
//            if (element.modifiers.contains(Modifier.PRIVATE)) {
//                processingEnv.messager
//                        .printMessage(Diagnostic.Kind.ERROR,
//                                "Binding annotation can not applied to private fields or methods.", element)
//            }

            if (annotation == BindPref::class.java || annotation == BindPrefString::class.java) {
                checkPreferenceAnnotation(element)
            }

            val targetPrefFragment = element.enclosingElement as TypeElement
            if (bindingMap.containsKey(targetPrefFragment)) {
                bindingMap[targetPrefFragment]?.add(element)
            } else {
                val fields = LinkedHashSet<Element>()
                fields.add(element)
                bindingMap[targetPrefFragment] = fields
            }
        }
    }

    /**
     * Checks whether a field is annotated correctly with [prefpin.BindPref] and prints error
     * message for incorrect one.
     */
    private fun checkPreferenceAnnotation(element: Element): Boolean {
        return if (isSubtypeOfType(element.asType(), "androidx.preference.Preference")) {
            true
        } else {
            processingEnv.messager
                    .printMessage(Diagnostic.Kind.ERROR,
                            "@PrefPin must be applied to Preference or its subclass fields", element)
            println("Bad Annotation")
            false
        }
    }

    @Throws(IOException::class)
    private fun writeBinding(targetClassName: String, annotationFields: Set<Element>) {
        var packageName: String? = null
        val lastDot = targetClassName.lastIndexOf('.')
        if (lastDot > 0) {
            packageName = targetClassName.substring(0, lastDot)
        }

        val targetSimpleClassName = targetClassName.substring(lastDot + 1)
        val bindingClassName = targetClassName + BINDING_CLASS_NAME_POSTFIX
        val bindingSimpleClassName = bindingClassName.substring(lastDot + 1)

        val targetClass = ClassName(packageName!!, targetSimpleClassName)

        val binding = TypeSpec.classBuilder(bindingSimpleClassName)
                .addModifiers(KModifier.PUBLIC)
                .addFunction(buildConstructor(targetClass, annotationFields))
                .build()

        val file = FileSpec.builder(packageName, bindingSimpleClassName)
                .addType(binding)
                .build()

        file.writeTo(processingEnv.filer)
    }

    private fun buildConstructor(targetClass: ClassName, annotationFields: Set<Element>): FunSpec {
        val constructorBuilder = FunSpec.constructorBuilder()
                .addAnnotation(UI_THREAD)
                .addModifiers(KModifier.PUBLIC)
                .addParameter("target", targetClass)

        for (element in annotationFields) {
            buildFieldBinding(constructorBuilder, element)
            buildOnClickBinding(constructorBuilder, element)
            buildOnChangeBinding(constructorBuilder, element)
        }

        return constructorBuilder.build()
    }

    private fun buildFieldBinding(constructorBuilder: FunSpec.Builder, element: Element) {
        val bindPref = element.getAnnotation(BindPref::class.java)
        if (bindPref != null) {
            val resourceId = bindPref.value

            constructorBuilder.addStatement(
                    "target.%L = target.findPreference<%T>(target.getString(%L)) as %T",
                    element.simpleName,
                    element.asType(),
                    resourceId,
                    element.asType()
            )
        }

        val bindPrefString = element.getAnnotation(BindPrefString::class.java)
        if (bindPrefString != null) {
            val key = bindPrefString.value

            constructorBuilder.addStatement(
                    "target.%L = target.findPreference<%T>(\"%L\") as %T",
                    element.simpleName,
                    element.asType(),
                    key,
                    element.asType()
            )
        }
    }

    private fun buildOnClickBinding(constructorBuilder: FunSpec.Builder, element: Element) {
        val onPrefClick = element.getAnnotation(OnPrefClick::class.java)
        if (onPrefClick != null) {
            val resourceIds = onPrefClick.value

            for (resourceId in resourceIds) {
                constructorBuilder.addStatement(
                        "target.findPreference<%T>(target.getString(%L))" +
                        "!!.setOnPreferenceClickListener(object : %T {\n" +
                                "override fun onPreferenceClick(preference: %T): Boolean {\n" +
                                        "\t\ttarget.%L(preference)\n" +
                                        "\t\treturn true\n" +
                                "\t}\n" +
                        "})", PREFERENCE, resourceId, CLICK_LISTENER, PREFERENCE, element.simpleName
                )
            }
        }

        val onPrefClickString = element.getAnnotation(OnPrefClickString::class.java)
        if (onPrefClickString != null) {
            val keys = onPrefClickString.value

            for (key in keys) {
                constructorBuilder.addStatement(
                        "target.findPreference<%T>(\"%L\")" +
                        "!!.setOnPreferenceClickListener(object : %T {\n" +
                                "override fun onPreferenceClick(preference: %T): Boolean {\n" +
                                "\t\ttarget.%L(preference)\n" +
                                "\t\treturn true\n" +
                                "\t}\n" +
                                "})", PREFERENCE, key, CLICK_LISTENER, PREFERENCE, element.simpleName
                )
            }
        }
    }

    private fun buildOnChangeBinding(constructorBuilder: FunSpec.Builder, element: Element) {
        val onPrefChange = element.getAnnotation(OnPrefChange::class.java)
        if (onPrefChange != null) {
            val resourceIds = onPrefChange.value

            for (resourceId in resourceIds) {
                constructorBuilder.addStatement(
                        "target.findPreference<%T>(target.getString(%L))" +
                        "!!.setOnPreferenceChangeListener(object : %T {\n" +
                                "override fun onPreferenceChange(preference: %T, newValue: Any): Boolean {\n" +
                                        "\t\ttarget.%L(preference, newValue)\n" +
                                        "\t\treturn true\n" +
                                "\t}\n" +
                        "})", PREFERENCE, resourceId, CHANGE_LISTENER, PREFERENCE, element.simpleName
                )
            }
        }

        val onPrefChangeString = element.getAnnotation(OnPrefChangeString::class.java)
        if (onPrefChangeString != null) {
            val keys = onPrefChangeString.value

            for (key in keys) {
                constructorBuilder.addStatement(
                        "target.findPreference<%T>(\"%L\")" +
                        "!!.setOnPreferenceChangeListener(object : %T {\n" +
                                "override fun onPreferenceChange(preference: %T, newValue: Any): Boolean {\n" +
                                "\t\ttarget.%L(preference, newValue)\n" +
                                "\t\treturn true\n" +
                                "\t}\n" +
                                "})", PREFERENCE, key, CHANGE_LISTENER, PREFERENCE, element.simpleName
                )
            }
        }
    }

    private fun isSubtypeOfType(typeMirror: TypeMirror, otherType: String): Boolean {
        if (isTypeEqual(typeMirror, otherType)) {
            return true
        }
        if (typeMirror.kind != TypeKind.DECLARED) {
            return false
        }
        val declaredType = typeMirror as DeclaredType
        val typeArguments = declaredType.typeArguments
        if (typeArguments.size > 0) {
            val typeString = StringBuilder(declaredType.asElement().toString())
            typeString.append('<')
            for (i in typeArguments.indices) {
                if (i > 0) {
                    typeString.append(',')
                }
                typeString.append('?')
            }
            typeString.append('>')
            if (typeString.toString() == otherType) {
                return true
            }
        }
        val element = declaredType.asElement() as? TypeElement ?: return false
        val superType = element.superclass
        if (isSubtypeOfType(superType, otherType)) {
            return true
        }
        for (interfaceType in element.interfaces) {
            if (isSubtypeOfType(interfaceType, otherType)) {
                return true
            }
        }
        return false
    }

    companion object {
        const val BINDING_CLASS_NAME_POSTFIX = "_PrefBinding"
        private val PREFERENCE = ClassName("androidx.preference", "Preference")
        private val CLICK_LISTENER = ClassName("androidx.preference.Preference", "OnPreferenceClickListener")
        private val CHANGE_LISTENER = ClassName("androidx.preference.Preference", "OnPreferenceChangeListener")
        private val UI_THREAD = ClassName("androidx.annotation", "UiThread")

        private fun isTypeEqual(typeMirror: TypeMirror, otherType: String): Boolean {
            return otherType == typeMirror.toString()
        }
    }
}
